projects/components/src/viewer/code-viewer/code-viewer.component.ts

```diff
@@ -145,14 +159,66 @@ export class CodeViewerComponent implements AfterViewInit, OnChanges, OnDestroy
     this.domMutationObserver.disconnect();
   }
 
+  public get isCodeEmpty(): boolean {
+    return isEmpty(this.code);
+  }
+
   public isLineHighlighted(lineNum: number): boolean {
     return (
       !isEmpty(this.highlightText) && this.codeLines[lineNum].toLowerCase().includes(this.highlightText.toLowerCase())
     );
   }
 
-  public onSearch(searchText: string): void {
-    this.searchText = searchText;
+  public onSearch(searchString: string): void {
+    // For case-insensitive search
+    const searchText = searchString.toLowerCase();
+    const codeTexts = this.codeTexts.map(text => text.toLowerCase());
+
+    // Remove existing child background elements
+    this.codeLineBackgroundElements.forEach(codeLineBackgroundElement => {
+      const codeLineBackgroundElemNode: Node = codeLineBackgroundElement.nativeElement;
+
+      while (codeLineBackgroundElemNode.firstChild) {
+        codeLineBackgroundElemNode.removeChild(codeLineBackgroundElemNode.firstChild);
+      }
+    });
+
+    const searchLen: number = searchText.length;
+    if (searchLen === 0) {
+      return;
+    }
+
+    const searchedPositions: Position[][] = Array.from(Array(codeTexts.length), () => new Array(0));
+
+    // Get all searched text positions
+    codeTexts.forEach((codeText, index) => {
+      for (let i = 0; i <= codeText.length - searchLen; i++) {
+        if (codeText.substr(i, searchLen) === searchText) {
+          searchedPositions[index].push({ start: i, end: i + searchLen });
+          i = i + searchLen - 1;
+        }
+      }
+    });
+
+    // Add background elements for searched positions
+    this.codeLineBackgroundElements.forEach((codeLineBackgroundElement, index) => {
+      searchedPositions[index].forEach(searchedPosition =>
+        (codeLineBackgroundElement.nativeElement as Node).appendChild(this.getBackgroundElement(searchedPosition))
+      );
```

Comment:
this is inefficient.

Instead, create a document fragment with all the new Elements inside it and add the fragment to the DOM in one go.

Reply:
So basically we can only create a fragment for each line, that's what I have done.

Reply:
we can create a fragment with all the children in it and then add it to dom.